src/main/java/com/learning/browsers/Browser.java

```diff
@@ -1,47 +1,58 @@
 package com.learning.browsers;
 
-import com.learning.configuration.PropertiesReader;
-import javax.annotation.PreDestroy;
+import java.util.Arrays;
+import java.util.function.Function;
+import org.openqa.selenium.JavascriptExecutor;
 import org.openqa.selenium.WebDriver;
-import org.openqa.selenium.chrome.ChromeDriver;
-import org.openqa.selenium.edge.EdgeDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.pagefactory.DefaultElementLocatorFactory;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
 
-public class Browser {
+@Component
+public class Browser implements DisposableBean {
 
-  private static final String BROWSER_PROPERTIES_NAME = "browser";
-  private static final String PATH_TO_EDGE_DRIVER = "pathToEdgeDriver";
-  private final PropertiesReader propertiesReader;
-  private WebDriver webDriver;
+  private final WebDriver webDriver;
 
-  public Browser(PropertiesReader propertiesReader) {
-    this.propertiesReader = propertiesReader;
-    this.webDriver = generateWebDriver();
+  public Browser(@Autowired WebDriverFactory webDriverFactory) {
+    this.webDriver = webDriverFactory.getWebDriver();
   }
 
-  public WebDriver getWebDriver() {
-    return webDriver;
-  }
-
-  private WebDriver generateWebDriver() {
-    String browserName = propertiesReader.getProperty(BROWSER_PROPERTIES_NAME);
-    if (browserName.equalsIgnoreCase("Chrome")) {
-      return new ChromeDriver();
-    } else if (browserName.equalsIgnoreCase("Edge")) {
-      System.setProperty("webdriver.edge.driver", propertiesReader.getProperty(PATH_TO_EDGE_DRIVER));
-      return new EdgeDriver();
-    } else {
-      throw new BrowserNotSupportedException("Unknown browser.");
-    }
+  public DefaultElementLocatorFactory getElementLocator() {
+    return new DefaultElementLocatorFactory(webDriver);
   }
 
   public String getCurrentPageURL() {
     return webDriver.getCurrentUrl();
   }
 
-  @PreDestroy
-  public void closeBrowser() {
-    webDriver.close();
+  public void open(String websiteURL, Loadable page) {
+    webDriver.get(websiteURL);
+    waitForPageToLoad(page);
+  }
+
+  public boolean waitForElementsToLoad(WebElement... elements) {
+    WebDriverWait wait = new WebDriverWait(webDriver, 10);
+    return wait.until(driver -> Arrays.stream(elements)
+        .allMatch(WebElement::isEnabled));
+  }
+
+  private void waitForPageToLoad(Loadable page) {
```

Comment:
We could return `boolean` here but if page fails to load there will be exception thrown. 
I don't see any usage for it here that is why I left void.

Reply:
This is actually a bad practice. Exception driven processing is not the way to go because:
- throwing exception won't allow you to perform cleanup steps (you can leave test environment is a state that will cause proceeding tests to fail if proper setup is not assured)
- not loading a page is not always a reason to break test execution (for example, you can implement a mechanism that will refresh page if one is not loaded, this is useful when you know that test environment performance is "sub-optimal")
- throwing exception does not give you a chance to properly log what has happened, thus performing RCA based on test execution log is very limited